enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,426 @@
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to allow operators managed by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) to communicate complex and potentially dangerous conditions and states to OLM. OLM will modify its behavior based on this information to protect the cluster.
+
+## Motivation
+
+OLM currently infers the state of an operator by inspecting the Kubernetes resources that define the operator. OLM does not provide operators with the means to communicate complex states that cannot be inferred without operator specific knowledge.
+
+Operators managed by OLM would benefit from the ability to communicate their state directly to OLM through a dedicated channel. Operators could communicate complex conditions and influence how OLM manages the operator, dramatically improving user experience. For example, an operator could communicate to OLM that the operator should not be upgraded or uninstalled while a critical process is being executed. Consider the following scenario based on OLM today:
+
+>An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM can provide both operator authors and cluster admins a significantly better user experience by creating a channel for the operator to communicate state that OLM cannot infer otherwise.
+
+### Goals
+
+- Provide operators with the means to communicate complex states that OLM cannot infer.
+- Provide operators with the means to inform OLM that the operator should not be upgraded or uninstalled.
+- Update OLM not to upgrade operators if the operator communicates that it should not be upgraded.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Support operator version rollback.
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate complex states to OLM that it cannot infer otherwise.
+
+#### Story 2 (core requirement)
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+### The OperatorStatus CRD
+
+OLM will introduce a new namespaced [CustomResourceDefinition](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/) called the OperatorStatus CRD. The OperatorStatus CRD will be used by operator to:
```

Comment:
We have the `Operator` CRD, why create a new CRD?

Reply:
Moved the new CRD approach to an alternative option.